ZoF support

sonicaj opened the pull request:
This PR aims to bring changes to py-libzfs so that we support ZoL(oF) as well. The changes would be split on two levels:

1) Autoconf
2) Cython

Ticket: #78123

sonicaj committed:
Initial commit for ZoF support

This commit sets up the basic CFLAGS and LDFLAGS which will be required for a working py-libzfs over zof.
Ticket: #78123

sonicaj committed:
Resolve function definition conflicts

This commit resolves a function definition conflict and also separates header check for libzfs.h as zfs_ioctl.h is not required by zof libzfs.h.

sonicaj committed:
Zfs_ioctl checks

This commit adds zfs_ioctl header checks so that we can tell if the header exists or not. For ZoF, we don't use it in libzfs, so this is necessary in that regard.

sonicaj committed:
Cover zfs_ioctl header usages

This commit covers zfs_ioctl header usages making sure that we don't use the header directly with zof as this is implemented differently in zof.

sonicaj committed:
Explicitly check for nvpair header location

This commit introduces checks for nvpair header location in autoconf as it differs under zof.

sonicaj committed:
Use correct path for header file ( nvpair.h )

This commit introduces changes to make sure we use the correct path in our pxd file for nvpair header.

sonicaj committed:
Zpool get status function signature check

This commit adds a check for zpool_get_status function signature as the signature differs in zof.

sonicaj committed:
Zpool_errata_t enum check

This commit adds capability to check zpool_errata_t enum in autoconf so that we can cover it's usage accordingly in the code.

sonicaj committed:
Correct zpool_get_status function signature

This commit corrects zpool_get_status function signature for zof.

sonicaj committed:
Check zpool_read_label function existence

This commit adds support for correctly configuring zpool_read_label signature and location as they differ in zof.

sonicaj committed:
Resolve zpool_read_label function conflicts

This commit resolves conflicts with the function signature of zpool_read_label and the location of the function as well.

sonicaj committed:
Check for ZPOOL_STATUS_NON_NATIVE_ASHIFT

This commit adds check for ZPOOL_STATUS_NON_NATIVE_ASHIFT enum as it does not exist in zof.

sonicaj committed:
Check ZPOOL_STATUS_NON_NATIVE_ASHIFT existence

This commit follows up on autoconf making sure that we use HAVE_ZPOOL_STATUS_NON_NATIVE_ASHIFT enum only if it exists.

sonicaj committed:
Check VDEV_AUX_ASHIFT_TOO_BIG existence in autoconf

This commit adds a check in autoconf to see if VDEV_AUX_ASHIFT_TOO_BIG exists before we use it as it is not present in zof.

sonicaj committed:
Check VDEV_AUX_ASHIFT_TOO_BIG existence in cython code

This commit makes sure we use VDEV_AUX_ASHIFT_TOO_BIG only if it is present.

sonicaj committed:
Check renameflags_t existence

This commit adds checks to see if renameflags_t exists in libzfs header.

sonicaj committed:
Cover zfs_rename and renameflags_t usages

As zof handles zfs_rename differently with different signature, this commit fixes that usages.

sonicaj committed:
Turn implicit function declarations into errors

This commit forces implicit function declaration warnings into errors making sure that we enforce stricter checks which will be particularly helpful for functions we have defined in libzutil header which are normally present in libzfs header in freebsd.

sonicaj committed:
Check libzutil header existence

This commit adds checks for libzutil header as it will be used in future for different function signatures as zof handles those functions differently.

sonicaj committed:
Import libzutil header if it exists

This commit imports libzutil header only if it exists. This header will be used for different varying function signatures which are present in libzfs for freebsd zfs.

sonicaj committed:
Check zpool_search_import usages

This commit adds checks to zpool_search_import function existence and signature.

sonicaj committed:
Cover usages of zpool_search_import

This commit covers all usages of zpool_search_import function to accommodate the different implementation of zof.

sonicaj committed:
Check zfs_prop_valid_for_type function signature

This commit adds checks for zfs_prop_valid_for_type function signature as it differs for zof.

sonicaj committed:
Cover zfs_prop_valid_for_type usages

This commit covers all usages of zfs_prop_valid_for_type to account for it's varying signature.

sonicaj committed:
Remove sys/vdev_impl.h header

This commit removes vdev_impl.h header file imports and it's usages as we no longer use them and also the import brings in a cascade of issues with zof as it imports certain files which start conflicting with the base system ( this issue is still in progress and will be resolved soon ).

sonicaj committed:
ZoF parameter allowed

This commit introduces a zof parameter which can be passed to the configure script and autoconf will generate a configure script whill force zof evaluation. It also normalizes spaces making them consistent which helps readability.

sonicaj committed:
Add config for setuptools

This commit adds a config file which will be populated by configure script. It will contain zfs type ( freebsd, zol ), and other configuration necessary for a successful build which setuptools will use to pass on and build a functioning libzfs.

william-gr commented on configure.ac:
```diff
@@ -11,46 +12,64 @@ define(SECTION_TITLE,
+AC_ARG_VAR(zof, AS_HELP_STRING([Force configure to run ZoL(f) checks]))
+AC_ARG_VAR(freebsd_src, AS_HELP_STRING([Provide path for freebsd source tree. This defaults to /usr/src/]))
+
+if [[ -f "/usr/local/include/libzfs/libzfs.h" ]] || [[ "$zof" = true ]]; then
```
Lets not hardcode /usr/local anywhere here, please.

Lets use an optional --prefix in configure script, defaulting to /usr/local

There are no guarantees everyone uses /usr/local, plus having it configurable makes it easier to support ZoL.

william-gr commented on libzfs.pyx:
```diff
@@ -2411,23 +2433,41 @@ cdef class ZFSObject(object):
+
+    ELSE:
+
+        def rename(self, new_name, force_unmount=False, recursive=False):
```
force_unmount is different from the other function define.

How about we use the same method scope but raise a RuntimeError if `nounmount` is used without HAVE_RENAMEFLAGS_T ?

sonicaj replied:
@william-gr i did that initially but then reverted to this. My reasoning at the time was that the function signature would show that we still accept nounmount ( even if we raise an exception ) - so i reverted to this case. Do you think that is okay and i can move back to having the same signature and raising an exception ?

william-gr replied:
Yes, I think that makes things simpler in the long run.

sonicaj replied:
Perfect, updating asap

william-gr commented on libzfs.pyx:
```diff
@@ -2705,22 +2745,23 @@ cdef class ZFSDataset(ZFSObject):
-    def get_send_progress(self, fd):
-        cdef zfs.zfs_cmd_t cmd
-        cdef int ret
+    IF HAVE_ZFS_IOCTL_HEADER:
```
I think I would rather have this method always defined but raise a NotImplementedError if not available.

william-gr reviewed (changes requested):
Nice job, I have just a few concerns.

sonicaj committed:
Make default header value configurable

This commit makes default header value configurable for the end user which can be passed to the configure script with --prefix=somevalue. It defaults to /usr/local right now.

sonicaj committed:
Update configure script

This commit updates configure script to reflect the latest changes introduced to autoconf.

william-gr commented on configure.ac:
```diff
@@ -11,46 +12,67 @@ define(SECTION_TITLE,
+AC_ARG_VAR(freebsd_src, AS_HELP_STRING([Provide path for freebsd source tree. This defaults to /usr/src/]))
+AC_ARG_VAR(zof_src, AS_HELP_STRING([Provide path for freebsd source tree. This defaults to /usr/local/]))
+
+zof_src=${zof_src:=/usr/local}
```
But thats not zof_src, thats PREFIX,  to where headers are installed. See previous comment about --prefix in configure, thats common practice.
